substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeBootImage.java

```diff
@@ -952,4 +973,299 @@ protected NativeTextSectionImpl(RelocatableBuffer relocatableBuffer, ObjectFile
+
+                @Override
+                public boolean hasNext() {
+                    return sourceIterator.hasNext();
+                }
+
+                @Override
+                public DebugLineInfo next() {
+                    return new NativeImageDebugLineInfo(sourceIterator.next());
+                }
+            };
+        }
+
+        public int getFrameSize() {
+            return compilation.getTotalFrameSize();
+        }
+
+        public List<DebugFrameSizeChange> getFrameSizeChanges() {
+            List<DebugFrameSizeChange> frameSizeChanges = new LinkedList<>();
+            for (Mark mark : compilation.getMarks()) {
+                // we only need to observe stack increment or decrement points
+                if (mark.id.equals("PROLOGUE_DECD_RSP")) {
+                    NativeImageDebugFrameSizeChange sizeChange = new NativeImageDebugFrameSizeChange(mark.pcOffset, EXTEND);
+                    frameSizeChanges.add(sizeChange);
+                    // } else if (mark.id.equals("PROLOGUE_END")) {
+                    // can ignore these
+                    // } else if (mark.id.equals("EPILOGUE_START")) {
+                    // can ignore these
+                } else if (mark.id.equals("EPILOGUE_INCD_RSP")) {
+                    NativeImageDebugFrameSizeChange sizeChange = new NativeImageDebugFrameSizeChange(mark.pcOffset, CONTRACT);
+                    frameSizeChanges.add(sizeChange);
+                    // } else if(mark.id.equals("EPILOGUE_END")) {
+                }
+            }
+            return frameSizeChanges;
+        }
+    }
+
+    /**
+     * implementation of the DebugLineInfo API interface
+     * that allows line number info to be passed to an
+     * ObjectFile when generation of debug info is enabled.
+     */
+    private class NativeImageDebugLineInfo implements DebugLineInfo {
+        private final int bci;
+        private final ResolvedJavaMethod method;
+        private final int lo;
+        private final int hi;
+
+        NativeImageDebugLineInfo(SourceMapping sourceMapping) {
+            NodeSourcePosition position = sourceMapping.getSourcePosition();
+            int bci = position.getBCI();
+            this.bci = (bci >= 0 ? bci : 0);
+            this.method = position.getMethod();
+            this.lo = sourceMapping.getStartOffset();
+            this.hi = sourceMapping.getEndOffset();
+        }
+
+        @Override
+        public String fileName() {
+            String name = className();
+            int idx = name.indexOf('$');
+            if (idx == 0) {
+                // name is $XXX so cannot associate with a file
+                return "";
+            }
+            if (idx > 0) {
+                // name is XXX$YYY so use outer class to derive file name
+                name = name.substring(0, idx);
+            }
+            return name.replace('.', '/') + ".java";
```

Comment:
Sometimes the source file isn't actually a `.java` file.  Is there any way to get access to the source name that was specified in the original `.class` file?  Or is this code only used when there is no debug info available?

Reply:
That's what is going on here. The call to declaringClass.getSourceFileName() actually reads the value from the class file -- well, strictly it finds the underlying Java class and reads the name out of the constant poll associated with the corresponding JVM InstanceKlass. So the code that constructs a name with extension java is just a fallback for when the bytecode contains no such entry.

Reply:
Ah, understood.  Thanks!

Reply:
This has been reworked but the new code is still using the classfile info where it exists and defaulting to append .java to the class name (or its initial segment up to $) otherwise.